src/NuGet.Core/NuGet.Protocol/Resources/ServiceIndexResourceV3.cs

```diff
@@ -105,7 +107,22 @@ private IReadOnlyList<ServiceIndexEntry> GetBestVersionMatchForType(NuGetVersion
             else
             {
                 // Find all entries with the same version.
-                return entries.Where(e => e.ClientVersion == bestMatch.ClientVersion).ToList();
+                List<ServiceIndexEntry> matchingEntries = new List<ServiceIndexEntry>();
+
+                foreach (var entry in entries)
+                {
+                    if (entry.ClientVersion == bestMatch.ClientVersion)
+                    {
+                        if (entry.Uri.Scheme == Uri.UriSchemeHttp && entry.Uri.Scheme != Uri.UriSchemeHttps && !_allowInsecureConnections)
+                        {
+                            throw new ArgumentException(string.Format(CultureInfo.CurrentCulture, Strings.Error_HttpServiceIndexUsage, entry.Uri));
```

Comment:
There is some context around why we didn't add warnings for HTTP resources in a feed in the past. It is documented [here](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1510220).

I am curious to understand if we perform this check once per every resource in the feed or once per resource and per package. If it is once per every resource in the feed, we should be okay I guess but if it is once per resource and per package then it may impact restore performance.